# Limpieza de Datos

In [22]:
import pandas as pd
from utils import convert_file

# Importar Datasets

### Dataset Movimientos-20190716

#### Importar el dataset y observar los primeros registros con head y verificar los tipos de datos con el método info.

In [23]:
movimientos_col_names = [
    'idb', 'id_producto', 'fecha_trans', 'stock_unidades', 'venta_unidades', 'precio_unitario',
    'cd_abast', 'is_prod_rol_local'
]

In [24]:
path = './data/VMI_Movimientos-20190716.txt'
movimientos_df = pd.read_csv(path,
                 delimiter = '\t',
                 header=None,
                 names=movimientos_col_names)

In [25]:
movimientos_df.head()

idb  id_producto  fecha_trans  stock_unidades  venta_unidades  \
0  242       228217     20160715            24.0             0.0   
1  242       228217     20160716            24.0             0.0   
2  242       228217     20160717            24.0             0.0   
3  242       228217     20160718            24.0             0.0   
4  242       228217     20160719            24.0             0.0   

   precio_unitario  cd_abast is_prod_rol_local  
0            13.63         0                SI  
1            13.63         0                SI  
2            13.63         0                SI  
3            13.63         0                SI  
4            13.63         0                SI

In [26]:
movimientos_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8347817 entries, 0 to 8347816
Data columns (total 8 columns):
 #   Column             Dtype  
---  ------             -----  
 0   idb                int64  
 1   id_producto        int64  
 2   fecha_trans        int64  
 3   stock_unidades     float64
 4   venta_unidades     float64
 5   precio_unitario    float64
 6   cd_abast           int64  
 7   is_prod_rol_local  object 
dtypes: float64(3), int64(4), object(1)
memory usage: 509.5+ MB


#### Ajustes al dataframe
- Convertir la columna fecha_trans a tipo de dato datetime.
- Convertir la columna is_prod_rol_local a tipo de dato string.
- Verificar por valores nulos y repetidos.
- Convertir la columna is_prod_rol_local a variable numerica donde 1 corresponde a SI y 0 corresponde a NO.
- Obtener las fechas máximas y minimas de transaciones
- Obtener una análisis descriptivo de los datos.

In [29]:
movimientos_df = convert_file(movimientos_df, 'fecha_trans', 'date')

TypeError: convert_file() missing 1 required positional argument: 'data_type'

In [ ]:
movimientos_df.isnull().sum(axis=0)

In [ ]:
movimientos_df.duplicated().sum()

In [ ]:
movimientos_df = convert_file(movimientos_df, 'is_prod_rol_local', 'str')

In [ ]:
movimientos_df['is_prod_rol_local'] = movimientos_df['is_prod_rol_local'].str.lower().apply(lambda x: 1 if x == 'si' else 0)

In [ ]:
max_date = movimientos_df['fecha_trans'].max()
min_date = movimientos_df['fecha_trans'].min()

print(f"Minimum date: {min_date}")
print(f"Maximum date: {max_date}")

In [ ]:
movimientos_df.info()

In [ ]:
movimientos_df.describe()

In [ ]:
# Verificar cuantos valores negativos tenemos en la columna venta_unidades
negative_values = movimientos_df[movimientos_df['venta_unidades'] < 0]
negative_values

#### Observaciones dataset Movimientos-20190716

1. El dataset contiene la información referente a los movimientos de productos comprendidos entre Julio 15 de 2016 hasta el día 14 de Julio de 2019.
2. El dataset contiene 8347817 registros y 8 columnas identificadas así:
    - idb: Identificador interno de Boca
    - id_producto: Identificador interno de Artículo
    - fecha: Fecha de la Transacción,  1 dia menos que la fecha actual
    - stock_unidades: Stock en unidades del Artículo en la Boca
    - venta_unidades: Venta en unidades
    - precio_unitario: Precio de venta unitario
    - cd_abast: Que CD de la region del local abastece este producto al local para productos almacenados.
    - is_prod_rol_local: SI= sigue en Rol, NO= no esta en Rol el producto en el local.
3. No se observan valores NAN ó valores repetidos.
4. El valor medio de inventario en unidades es de 20.65585 unidades.
5. El valor medio de ventas en unidades es de 0.3133696 unidades.
6. La media del precio unitario es de 64.26050 pesos.

### Dataset Promociones-20190715

#### Importar el dataset y observar los primeros registros con head y verificar los tipos de datos con el método info.

In [ ]:
promo_col_names = [
        'idb', 'id_producto', 'fecha_desde', 'fecha_hasta', 'ventas_estimadas_und',
        'nro_prom', 'fecha_trans', 'porcentaje_dcto'
]

In [ ]:
promo_df = pd.read_csv("./data/VMI_Promociones-20190715.txt",
                       encoding="utf-8",
                       header=None,
                       delimiter='\t',
                       names=promo_col_names)

In [ ]:
promo_df.head()

In [ ]:
promo_df.info()

#### Ajustes al dataframe
- Convertir la columnas fecha_desde, fecha_hasta y fecha_trans a tipo de dato datetime.
- Verificar si existen valores nulos y/o repetidos.
- Obtener las fechas máximas y minimas de transaciones.
- Crear una nueva columnas llamada duracion_promo para obtener el delta de tiempo entre la fecha_desde y la fecha_hasta de una promoción.
- Obtener una análisis descriptivo de los datos.

In [ ]:
promo_df = convert_file(promo_df, 'fecha_desde',  'date')

In [ ]:
promo_df = convert_file(promo_df, 'fecha_hasta',  'date')

In [ ]:
promo_df = convert_file(promo_df, 'fecha_trans', 'date')

In [ ]:
promo_df.isnull().sum(axis=0)

In [ ]:
promo_df.duplicated().sum()

In [ ]:
min_date = promo_df['fecha_desde'].min()
max_date = promo_df['fecha_hasta'].max()


print(f"Minimum date: {min_date}")
print(f"Maximum date: {max_date}")

In [ ]:
min_date = promo_df['fecha_trans'].min()
max_date = promo_df['fecha_trans'].max()

print(f"Minimum date: {min_date}")
print(f"Maximum date: {max_date}")

In [ ]:
# Crear una nueva columna llamada duracion de la promoción.
promo_df['duracion_promo'] = promo_df['fecha_hasta'] - promo_df['fecha_desde']

In [ ]:
promo_df.describe()

In [ ]:
promo_df.info()

#### Observaciones dataset Promociones-20190715

1. El dataset contiene la información referente a las promociones realizadas desde 2016-07-19 hasta el 2019-07-16.
2. El dataset contiene 285253 registros y 9 columnas descritas así:
    - idb:	Identificador interno de Boca
    - id_producto:	Identificador interno de Artículo
    - fecha-desde: 	Fecha desde de la vigencia de la Promoción
    - fecha_hasta:	Fecha hasta de la vigencia de la Promoción
    - ventas_estimadas_und: Estimado de ventas en la promoción en Unidades
    - nro_prom: Número de la Promoción
    - fecha_trans: Fecha de la Información
    - porcentaje_dcto: Porcentaje de Descuento
    - duracion_promo: La diferencia entre fecha_desde y fecha_hasta.
3. No se observan valores NAN ó duplicados.
4. La media de ventas estimadas es de 24.62 unidades.
5. El máximo porcentaje de descuento es de 18% y el mínimo es de 3%.
6. La duración máxima de una promoción es de 18 días.
7. La duración media de una promoción es de 8 días.

### Dataset ProvProdu-20190712

#### Importar el dataset y observar los primeros registros con head y verificar los tipos de datos con el método info.

In [ ]:
prov_prod_col_names = [
    'id_proveedor', 'razon_social', 'id_sector', 'desc_sector', 'id_seccion',
    'desc_seccion', 'id_gran_flia', 'desc_gran_flia', 'id_flia', 'desc_flia',
    'id_sub_flia', 'desc_sub_flia', 'id_articulo', 'id_producto',
    'desc_producto', 'id_region', 'fecha'
]

In [ ]:
prov_prod_df = pd.read_csv("./data/VMI_ProvProdu-20190712.txt",
                           encoding="ISO-8859-1",
                           header=None,
                           delimiter='\t',
                           names=prov_prod_col_names)

In [ ]:
prov_prod_df.head()

In [ ]:
prov_prod_df.info()

#### Ajustes al dataframe
- Arreglar errores de ortografia para las columnas desc_sector y desc_gran_flia.
- Convertir las columnas de tipo object a string y luego a lower case, esto nos va a facilitar realizar el análisis de variables categoricas.
- Verificar valores nulos y/o duplicados.
- Convertir la columna fecha a tipo datetime.
- Hacer un análisis descriptivo.

In [ ]:
prov_prod_df['desc_gran_flia'] = prov_prod_df['desc_gran_flia'].replace(to_replace=r'Ba¤o', value='Baño', regex=True)
prov_prod_df['desc_sector'] = prov_prod_df['desc_sector'].replace(to_replace=r'Almac‚n', value='Almacén', regex=True)

In [ ]:
columnas_categoricas = ['razon_social', 'desc_sector', 'desc_seccion', 'desc_gran_flia', 'desc_flia', 'desc_sub_flia', 'desc_producto']
prov_prod_df = convert_file(prov_prod_df, 'x', columnas_categoricas, 'str_m' )

In [ ]:
prov_prod_df = prov_prod_df.applymap(lambda x: x.lower() if isinstance(x, str) else x)

In [ ]:
prov_prod_df.info()

In [ ]:
prov_prod_df = convert_file(prov_prod_df, 'fecha', 'date')

### Dataset Locales-20190712

In [ ]:
locales_df = pd.read_csv("./data/VMI_Locales-20190712.txt",
                         encoding="ISO-8859-1",
                         header=None,
                         delimiter='\t')

In [ ]:
locales_col_names = [
    'id_cadena', 'desc_cadena', 'id_region', 'desc_region', 'idb',
    'desc_local', 'latitud', 'longitud', 'pd'
]
locales_df.columns = locales_col_names

In [ ]:
locales_df['desc_local'] = locales_df['desc_local'].replace(to_replace=r'Rinc¢n', value='Rincón', regex=True)

In [ ]:
locales_df.head()

In [ ]:
locales_df.info()

In [ ]:
locales_df.isnull().sum(axis=0)

In [ ]:
locales_df.describe()

# Análisis Descriptivo

In [ ]:
merged_df = pd.merge(prov_prod_df, locales_df, on='id_region')
merged_df.head()

In [ ]:
merged_df.info()

In [ ]:
merged_df = pd.merge(merged_df, promo_df, on="idb")

In [ ]:
merged_df.describe()

In [ ]:
print(merged_df.loc[:, ['id_producto_x', 'id_producto_y']])